README.md
# NAME

<div>
    <a href="https://travis-ci.org/thedatahub/Catmandu-Store-Datahub"><img src="https://travis-ci.org/thedatahub/Catmandu-Store-Datahub.svg?branch=master"></a>
</div>

Catmandu::Store::Datahub - Store/retrieve items from the Datahub

# SYNOPSIS

A module that allows to interface with the Datahub as a Catmandu::Store.

Supports retrieving, adding, deleting and updating of data.

# DESCRIPTION

Configure the [Datahub](https://github.com/thedatahub/Datahub) as a [store](http://librecat.org/Catmandu/#stores) for [Catmandu](http://librecat.org/).

With Catmandu, it is possible to convert (almost) any data to [LIDO](http://lido-schema.org/), which is suitable for importing in the Datahub. This module allows you to integrate the importing in your Catmandu workflow by setting up a Catmandu-compatible interface between the Datahub and Catmandu.

Note that you must convert your data to LIDO in order to use this module. All other formats will result in an error.

# CONFIGURATION

To configure the store, the location of the Datahub is required. As OAuth2 is used, a client id and secret are also required, as well as a username and a password.

- `url`

    base url of the Datahub (e.g. _http://www.datahub.be_).

- `client_id`

    OAuth2 client ID.

- `client_secret`

    OAuth2 client secret.

- `username`

    Datahub username.

- `password`

    Datahub password.

# USAGE

See [the Catmandu documentation](http://librecat.org/Catmandu/#stores) for more information on how to use Stores.

# SEE ALSO

[Catmandu::LIDO](https://metacpan.org/pod/Catmandu::LIDO) and [Catmandu](https://metacpan.org/pod/Catmandu)

# AUTHORS

Matthias Vandermaesen <matthias.vandermaesen@vlaamsekunstcollectie.be>
Pieter De Praetere <pieter@packed.be>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by PACKED, vzw, Vlaamse Kunstcollectie, vzw.

This is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License, Version 3, June 2007.


Changes
Revision history for Catmandu-Store-Datahub

{{$NEXT}}

0.08  2017-10-29 17:27:18 CET

    - Fixed: issue with Catmandu::HTTPError getting the wronge response params.

0.07  2017-10-29 15:08:28 CET

    - Changed: Configured dzil building.
    - Changed: licensing to GPLv3.

0.06  2017-07-08 13:55:03 CEST

    - fixed: missing ID now fails gracefully.

0.05  2017-06-08 10:57:50 CEST
    - fixed: LWP::Useragent breaks with 'select failed: invalid argument' error after processing multiple records
    - changed: confrom to new Catmandu::Store::Datahub API
    - changed: POST does not need to know the id of the enclosed entity per RFC 2616
    - changed: POST should not check if a resource exists a piori per RFC 2616
    - changed: consistent coding in API methods
    - changed: get the proper error message from the HTTP::Response headers
    - fixed: datahub requires Content-Type: application/lido+xml

0.04  2017-05-24 12:44:09 CEST
    - removed: references to lido
    - added: don't push a record to the datahub if it doesn't have an identifier

0.03  2017-05-19 10:36:27 CEST
    - API refactoring

0.02  2017-01-04 14:42:29 UTC
    - Added documentation

0.01  2016-12-08 12:32:11 CET
    - Initial release


lib/Catmandu/Store/Datahub.pm
package Catmandu::Store::Datahub;

our $VERSION = '0.07';

use Catmandu::Sane;

use Moo;
use Catmandu::Store::Datahub::Bag;
use Catmandu::Store::Datahub::OAuth;
use LWP::UserAgent;

with 'Catmandu::Store';

has url           => (is => 'ro', required => 1);
has client_id     => (is => 'ro', required => 1);
has client_secret => (is => 'ro', required => 1);
has username      => (is => 'ro', required => 1);
has password      => (is => 'ro', required => 1);

has client       => (is => 'lazy');
has access_token => (
    is      => 'lazy',
    writer  => '_set_access_token',
    builder => '_build_access_token'
);

##
# TODO: error reporting 'n stuff

sub _build_client {
    my $self = shift;
    return LWP::UserAgent->new();
}

sub _build_access_token {
    my $self = shift;
    return $self->generate_token();
}

sub set_access_token {
    my $self = shift;
    # Used to regenerate the token when it becomes invalid
    return $self->_set_access_token($self->generate_token());
}

sub generate_token {
    my $self = shift;
    my $oauth = Catmandu::Store::Datahub::OAuth->new(username => $self->username, password => $self->password, client_id => $self->client_id, client_secret => $self->client_secret, url => $self->url);
    return $oauth->token();
}

1;

=head1 NAME

=for html <a href="https://travis-ci.org/thedatahub/Catmandu-Store-Datahub"><img src="https://travis-ci.org/thedatahub/Catmandu-Store-Datahub.svg?branch=master"></a>

Catmandu::Store::Datahub - Store/retrieve items from the Datahub

=head1 SYNOPSIS

A module that allows to interface with the Datahub as a Catmandu::Store.

Supports retrieving, adding, deleting and updating of data.

=head1 DESCRIPTION

Configure the L<Datahub|https://github.com/thedatahub/Datahub> as a L<store|http://librecat.org/Catmandu/#stores> for L<Catmandu|http://librecat.org/>.

With Catmandu, it is possible to convert (almost) any data to L<LIDO|http://lido-schema.org/>, which is suitable for importing in the Datahub. This module allows you to integrate the importing in your Catmandu workflow by setting up a Catmandu-compatible interface between the Datahub and Catmandu.

Note that you must convert your data to LIDO in order to use this module. All other formats will result in an error.

=head1 CONFIGURATION

To configure the store, the location of the Datahub is required. As OAuth2 is used, a client id and secret are also required, as well as a username and a password.

=over

=item C<url>

base url of the Datahub (e.g. I<http://www.datahub.be>).

=item C<client_id>

OAuth2 client ID.

=item C<client_secret>

OAuth2 client secret.

=item C<username>

Datahub username.

=item C<password>

Datahub password.

=back

=head1 USAGE

See L<the Catmandu documentation|http://librecat.org/Catmandu/#stores> for more information on how to use Stores.

=head1 SEE ALSO

L<Catmandu::LIDO> and L<Catmandu>

=head1 AUTHORS

Matthias Vandermaesen <matthias.vandermaesen@vlaamsekunstcollectie.be>
Pieter De Praetere <pieter@packed.be>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2016 by PACKED, vzw, Vlaamse Kunstcollectie, vzw.

This is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License, Version 3, June 2007.

=cut


lib/Catmandu/Store/Datahub/API.pm
package Catmandu::Store::Datahub::API;

use strict;
use warnings;

use Catmandu;
use Moo;
use JSON;

use LWP::UserAgent;

has url           => (is => 'ro', required => 1);
has client_id     => (is => 'ro', required => 1);
has client_secret => (is => 'ro', required => 1);
has username      => (is => 'ro', required => 1);
has password      => (is => 'ro', required => 1);

has client       => (is => 'lazy');
has access_token => (
    is      => 'lazy',
    writer  => '_set_access_token',
    builder => '_build_access_token'
);

sub _build_client {
    my $self = shift;
    return LWP::UserAgent->new();
}

sub _build_access_token {
    my $self = shift;
    return $self->generate_token();
}

sub set_access_token {
    my $self = shift;
    # Used to regenerate the token when it becomes invalid
    return $self->_set_access_token($self->generate_token());
}

sub generate_token {
    my $self = shift;
    my $oauth = Catmandu::Store::Datahub::OAuth->new(username => $self->username, password => $self->password, client_id => $self->client_id, client_secret => $self->client_secret, url => $self->url);
    return $oauth->token();
}

sub get {
    my ($self, $id) = @_;
    my $url = sprintf('%s/api/v1/data/%s', $self->url, $id);

    my $response = $self->client->get($url, Authorization => sprintf('Bearer %s', $self->access_token));

    if ($response->is_success) {
        return decode_json($response->decoded_content);
    } elsif ($response->code == 401) {
        my $error = decode_json($response->decoded_content);
        if ($error->{'error_description'} eq 'The access token provided has expired.') {
            $self->set_access_token();
            return $self->get($id);
        }
    } elsif ($response->code == 404) {
        return {};
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->headers->header('message'),
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content,
            });
        return undef;
    }
}

sub add {
    my ($self, $data) = @_;
    my $url = sprintf('%s/api/v1/data.lidoxml', $self->url);

    my $token = $self->access_token;
    my $response;

    $response = $self->client->post($url, Content_Type => 'application/lido+xml', Authorization => sprintf('Bearer %s', $token), Content => $data);

    if ($response->is_success) {
        return $response->decoded_content;
    } elsif ($response->code == 401) {
        my $error = decode_json($response->decoded_content);
        if ($error->{'error_description'} eq 'The access token provided has expired.') {
            $self->set_access_token();
            return $self->add($data);
        }
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->headers->header('message'),
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content,
            });
        return undef;
    }
}

sub update {
    my ($self, $id, $data) = @_;
    my $url = sprintf('%s/api/v1/data/%s', $self->url, $id);

    my $token = $self->access_token;
    my $response;

    $response = $self->client->put($url, Content_Type => 'application/lido+xml', Authorization => sprintf('Bearer %s', $token), Content => $data);

    if ($response->is_success) {
        return $response->decoded_content;
    } elsif ($response->code == 401) {
        my $error = decode_json($response->decoded_content);
        if ($error->{'error_description'} eq 'The access token provided has expired.') {
            $self->set_access_token();
            return $self->update($id, $data);
        }
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->headers->header('message'),
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content,
            });
        return undef;
    }
}

sub delete {
    my ($self, $id) = @_;
    my $url = sprintf('%s/api/v1/data/%s', $self->url, $id);

    my $token = $self->access_token;
    my $response;

    $response = $self->client->delete($url, Authorization => sprintf('Bearer %s', $token));

    if ($response->is_success) {
        return $response->decoded_content;
    } elsif ($response->code == 401) {
        my $error = decode_json($response->decoded_content);
        if ($error->{'error_description'} eq 'The access token provided has expired.') {
            $self->set_access_token();
            return $self->delete($id);
        }
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->headers->header('message'),
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content,
            });
        return undef;
    }
}

sub list {
    my ($self) = @_;
    my $url = sprintf('%s/api/v1/data', $self->url);

    my $token = $self->access_token;
    my $response = $self->client->get($url, Authorization => sprintf('Bearer %s', $token));

    if ($response->is_success) {
        return decode_json($response->decoded_content);
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->headers->header('message'),
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content,
            });
        return undef;
    }
}

1;

__END__


lib/Catmandu/Store/Datahub/Bag.pm
package Catmandu::Store::Datahub::Bag;

use Moo;
use Scalar::Util qw(reftype);
use LWP::UserAgent;
use Catmandu::Util qw(is_string require_package);
use Time::HiRes qw(usleep);
use Catmandu::Sane;
use Catmandu::Store::Datahub::API;
use JSON;

with 'Catmandu::Bag';

has api => (is => 'lazy');

sub _build_api {
    my $self = shift;
    my $api = Catmandu::Store::Datahub::API->new(
        url           => $self->store->url,
        client_id     => $self->store->client_id,
        client_secret => $self->store->client_secret,
        username      => $self->store->username,
        password      => $self->store->password
    );
    return $api;
}

##
# before add in ::Bag creates a _id tag, which is useful for hashes and NoSQL-dbs, but breaks our
# XML conversion. You *can* remove it in your add/update function, but that feels unclean.
# You cannot use before add here to remove the _id, as this one is added before the before add
# from ::Bag (see http://search.cpan.org/~ether/Moose-2.1806/lib/Moose/Manual/MethodModifiers.pod).
# But around is called last. So we use it here.
around add => sub {
    my $orig = shift;
    my ($self, $data) = @_;
    delete $data->{'_id'};
    return $self->$orig($data);
};

sub generator {
    my $self = shift;
    # api/v1/data -> results ; not paginated
    my $stack = $self->api->list()->{'results'};
    return sub {
        return pop @{$stack};
    };
}

sub each {
    my ($self, $sub) = @_;
    my $n = 0;
    my $stack = $self->api->list()->{'results'};
    while (my $item = pop @{$stack}) {
        $sub->($item);
        $n++;
    }
    return $n;
}


##
# Return a record identified by $id
sub get {
    my ($self, $id) = @_;
    return $self->api->get($id);
}

##
# Create a new record
sub add {
    my ($self, $data) = @_;
    return $self->api->add($data->{'_'});
}

##
# Delete a record
sub delete {
    my ($self, $id) = @_;
    return $self->api->delete($id);
}

sub delete_all {}

1;

__END__


lib/Catmandu/Store/Datahub/OAuth.pm
package Catmandu::Store::Datahub::OAuth;

use LWP::UserAgent;
use JSON;
use Moo;

has username => (is => 'ro', required => 1);
has password => (is => 'ro', required => 1);

has client_id     => (is => 'ro', required => 1);
has client_secret => (is => 'ro', required => 1);

has ua => (is => 'lazy');


sub _build_ua {
    my $self = shift;
    return LWP::UserAgent->new();
}

sub token {
    my $self = shift;
    # Replace this with something better as soon as we figure out how we can get rid of the password
    return $self->get_token_u_p();
}

sub get_token_u_p {
    my ($self) = @_;
    my $auth_url = 'http://datahub.app/oauth/v2/token?grant_type=password&username=%s&password=%s&client_id=%s&client_secret=%s';
    my $response = $self->ua->get(sprintf($auth_url, $self->username, $self->password, $self->client_id, $self->client_secret));
    if ($response->is_success) {
        my $token_raw = $response->decoded_content;
        my $token_parsed = decode_json($token_raw);
        return $token_parsed->{'access_token'};
    } else {
        print($response->status_line."\n");
        print($response->decoded_content."\n");
        return undef;
    }
}



1;

lib/Catmandu/Store/Datahub/Generator.pm
package Catmandu::Store::Datahub::Generator;

use LWP::UserAgent;
use JSON;
use Moo;

use Catmandu::Sane;
use Catmandu::Store::Datahub::OAuth;

has token => (is => 'ro', required => 1);
has url   => (is => 'ro', required => 1);

has list  => (is => 'rw', default => sub {
    return [];
});

has ua   => (is => 'lazy');

sub _build_ua {
    my $self = shift;
    return LWP::UserAgent->new();
}

sub set_list {
    my ($self) = @_;
    my $url = sprintf('%s/api/v1/data', $self->url);
    my $response = $self->ua->get($url, Authorization => sprintf('Bearer %s', $self->token));
    if ($response->is_success) {
        my $json = $response->decoded_content;
        my $decoded_json = decode_json($json);
        $self->list = $decoded_json->['results'];
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->status_line,
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content,
            });
    }
}

sub get_single {
    my ($self, $id) = @_;
    my $url = sprintf('%s/api/v1/data/%s', $self->url, $id);
    my $response = $self->ua->get($url, Authorization => sprintf('Bearer %s', $self->token));
    if ($response->is_success) {
        return $response->decoded_content;
    } else {
        Catmandu::HTTPError->throw({
                code             => $response->code,
                message          => $response->status_line,
                url              => $response->request->uri,
                method           => $response->request->method,
                request_headers  => [],
                request_body     => $response->request->decoded_content,
                response_headers => [],
                response_body    => $response->decoded_content,
            });
        return undef;
    }
}

sub next {
    my ($self) = @_;
    my $next = shift @{$self->list};
    return $self->get_single($next->['data_pids'][0]);
}

1;

Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.010.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "Store/retrieve items from the Datahub",
  "dist_author" => [
    "Matthias Vandermaesen <matthias.vandermaesen\@vlaamsekunstcollectie.be>"
  ],
  "dist_name" => "Catmandu-Store-Datahub",
  "dist_version" => "0.07",
  "license" => "gpl",
  "module_name" => "Catmandu::Store::Datahub",
  "recursive_test_files" => 1,
  "requires" => {
    "Catmandu" => "1.0304",
    "JSON" => "2.90",
    "Lido::XML" => "0.07",
    "URL::Encode" => "0.03"
  },
  "test_requires" => {
    "Test::Deep" => "0.112",
    "Test::Exception" => "0.43",
    "Test::More" => "0.99",
    "Test::Pod" => 0
  }
);


my %fallback_build_requires = (
  "Module::Build" => "0.28",
  "Test::Deep" => "0.112",
  "Test::Exception" => "0.43",
  "Test::More" => "0.99",
  "Test::Pod" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);


$build->create_build_script;
